src/async_impl/response.rs

```diff
@@ -106,16 +121,53 @@ impl fmt::Debug for Response {
     }
 }
 
+impl decoder::PollChunk for Body {
+    fn poll_chunk(&mut self) -> io::Result<Option<Chunk>> {
+        match self.poll() {
+            Ok(Async::Ready(Some(chunk))) => Ok(Some(chunk)),
+            Ok(Async::Ready(None)) => Ok(None),
+            Ok(Async::NotReady) => Err(io::ErrorKind::WouldBlock.into()),
+            Err(e) => Err(::error::into_io(e)),
+        }
+    }
+}
+
+pub struct DecodedBody {
```

Comment:
I've just got a few questions (sorry I don't usually do this kind of programming):

- Would you expect the synchronous response to consume the `Decoder: Stream`, or the `DecoderImplWhateverNameNotSuperImportant: Read`? I guess the inner decoder could be both `Read` and `Stream`, so the plain text decoding just means passing on chunks
- If we're going to emit chunks for the gzip decoder, then we'll need to read them into some buffer. I'm not sure how big that should be or where it should come from

So I'm not sure whether this should change the way I've done the `Decoder` completely or if it's just a matter of the async `Response` using `DecoderImplWhateverNameNotSuperImportant: Stream` and the sync `Response` using `DecoderImplWhateverNameNotSuperImportant: Read`.